2.0: test suite is failing

```console
+ cd liburing-liburing-2.0
+ /usr/bin/make -O -j48 V=1 VERBOSE=1 runtests
make[1]: Nothing to be done for 'all'.
make[1]: Nothing to be done for 'all'.
make[1]: Nothing to be done for 'all'.
make[1]: Entering directory '/home/tkloczko/rpmbuild/BUILD/liburing-liburing-2.0/test'
Running test 232c93d07b74-test
Running test 35fa71a030ca-test
Test 35fa71a030ca-test timed out (may not be a failure)
Running test 500f9fbadef8-test
Polling not supported in current dir, test skipped
Running test 7ad0e4b2f83c-test
Running test 8a9973408177-test
Running test 917257daa0fe-test
Running test a0908ae19763-test
Running test a4c0b3decb33-test
Running test accept
SQPOLL skipped for regular user
Running test accept-link
Running test accept-reuse
Running test accept-test
Running test across-fork
Running test splice
Running test b19062a56726-test
Running test b5837bd5311d-test
Running test ce593a6c480a-test
Running test close-opath
Running test connect
cqe 1, res 0, wanted -125
test_connect_timeout(): failed
Test connect failed with ret 1
Running test cq-full
Running test cq-overflow
Running test cq-peek-batch
Running test cq-ready
Running test cq-size
Running test d4ae271dfaae-test
Test requires root, skipping
Running test d77a67ed5f27-test
SQPOLL skipped for regular user
Running test defer
ring setup failed
Test defer failed with ret 1
Running test double-poll-crash
timeout: the monitored command dumped core
./runtests.sh: line 66: 3299849 Segmentation fault      timeout -s INT -k $TIMEOUT $TIMEOUT ./$test_name $dev
Test double-poll-crash failed with ret 139
Running test eeed8b54e0df-test
Running test eventfd
Running test eventfd-disable
Running test eventfd-ring
Running test fadvise
Running test fallocate
Running test fc2a85cb02ef-test
Test needs failslab/fail_futex/fail_page_alloc enabled, skipped
Running test file-register
Skipping files not supported
Running test file-update
Running test files-exit-hang-poll
Running test files-exit-hang-timeout
Running test fixed-link
Running test fsync
Running test io-cancel
Test io-cancel timed out (may not be a failure)
Running test io_uring_enter
io_uring_queue_init: Cannot allocate memory
Test io_uring_enter failed with ret 1
Running test io_uring_register
RELIMIT_MEMLOCK: 65536 (65536)
io_uring_register(-1, 0, (nil), 0)
io_uring_register(3, 0, (nil), 0)
io_uring_register(4, 4294967295, (nil), 0)
io_uring_register(4, 0, 0x7ffee42b9e40, 1)
io_uring_register(4, 0, 0x7ffee42b9e40, 1)
io_uring_register(4, 0, 0x7ffee42b9e40, 1)
Unable to map a huge page.  Try increasing /proc/sys/vm/nr_hugepages by at least 1.
Skipping the hugepage test
reserve file-backed buffers
io_uring_register(4, 0, 0x7ffee42b9e40, 1)
io_uring_register of 131072 bytes failed with ENOMEM (expected).
io_uring_register of 65536 bytes failed with ENOMEM (expected).
successfully registered 32768 bytes (0).
io_uring_register(4, 0, 0x43b2c0, 1025)
io_uring_register(4, 0, 0x43b2c0, 1024)
expected success, got 12
allocated 17179869180 bytes of address space
io_uring_register(4, IORING_REGISTER_FILES, 0x7f56929fd000, 32767)...succeeded
io_uring_register(4, IORING_UNREGISTER_FILES, 0, 0)...succeeded
io_uring_submit:
        opcode: 6
        flags:  0x00000000
        fd:     4
        poll_events: 0x00000005
io_uring_register(4, 2, 0x7ffee42b9e40, 1)
FAIL
Test io_uring_register failed with ret 1
Running test io_uring_setup
io_uring_setup(0, 0x7ffda33ca170), flags: none, feat: none, resv: 0x00000000 0x00000000 0x00000000, sq_thread_cpu: 0
io_uring_setup(1, (nil)), flags: none, feat: none, resv: , sq_thread_cpu: 0
io_uring_setup(1, 0x7ffda33ca170), flags: none, feat: none, resv: 0x00000001 0x00000001 0x00000001, sq_thread_cpu: 0
io_uring_setup(1, 0x7ffda33ca170), flags: 0xffffffff, feat: none, resv: 0x00000000 0x00000000 0x00000000, sq_thread_cpu: 0
io_uring_setup(1, 0x7ffda33ca170), flags: IORING_SETUP_SQ_AFF, feat: none, resv: 0x00000000 0x00000000 0x00000000, sq_thread_cpu: 0
io_uring_setup(1, 0x7ffda33ca170), flags: IORING_SETUP_SQPOLL|IORING_SETUP_SQ_AFF, feat: none, resv: 0x00000000 0x00000000 0x00000000, sq_thread_cpu: 48
Needs root, not flagging as an error
PASS
Running test iopoll
File/device/fs doesn't support polled IO
Running test lfs-openat
Running test lfs-openat-write
Running test link
wrong CQE order, got 1, expected 0
test_link_fail_ordering last failed
Test link failed with ret 1
Running test link-timeout
Running test link_drain
Running test madvise
Running test nop
Running test nop-all-sizes
Running test open-close
Running test openat2
Running test personality
Not root, skipping
Running test pipe-eof
Running test pipe-reuse
Running test poll
Running test poll-cancel
Running test poll-cancel-ton
timeout: the monitored command dumped core
./runtests.sh: line 66: 3300353 Segmentation fault      timeout -s INT -k $TIMEOUT $TIMEOUT ./$test_name $dev
Test poll-cancel-ton failed with ret 139
Running test poll-link
Running test poll-many
timeout: the monitored command dumped core
./runtests.sh: line 66: 3300383 Segmentation fault      timeout -s INT -k $TIMEOUT $TIMEOUT ./$test_name $dev
Test poll-many failed with ret 139
Running test poll-mshot-update
timeout: the monitored command dumped core
./runtests.sh: line 66: 3300395 Segmentation fault      timeout -s INT -k $TIMEOUT $TIMEOUT ./$test_name $dev
Test poll-mshot-update failed with ret 139
Running test poll-ring
Running test poll-v-poll
Running test probe
Running test read-write
SQPOLL requires root, skipping
buffer reg failed: -12
test_io failed 0/0/0/1/0
Test read-write failed with ret 1
Running test register-restrictions
Running test rename
Running test ring-leak
Running test ring-leak2
Running test rw_merge_test
Running test self
Running test send_recv
SQPOLL skipped for regular user
SQPOLL skipped for regular user
Running test send_recvmsg
Running test shared-wq
Running test short-read
Running test shutdown
Running test sigfd-deadlock
Running test socket-rw
Running test socket-rw-eagain
Running test sq-full
Running test sq-poll-dup
queue_init: -1/0
test 0 0 failed
Test sq-poll-dup failed with ret 1
Running test sq-poll-kthread
sqpoll requires root!
sqpoll requires root!
test_sq_poll_kthread_stopped_exit: skipped
test_sq_poll_kthread_stopped_exit: skipped
test_sq_poll_kthread_stopped_noexit: skipped
Running test sq-poll-share
queue_init: -1/0
Test sq-poll-share failed with ret 1
Running test sqpoll-disable-exit
Running test sqpoll-exit-hang
./sqpoll-exit-hang: skipped, not root
Running test sqpoll-sleep
./sqpoll-sleep: skipped, not root
Running test sq-space_left
Running test stdout
This is a pipe test
This is a fixed pipe test
Running test submit-reuse
Running test teardowns
Running test thread-exit
Running test timeout
Running test timeout-new
ring_sq setup failed: -1
Test timeout-new failed with ret 1
Running test timeout-overflow
Skipping
Running test unlink
Running test wakeup-hang
Running test sendmsg_fs_cve
chroot(): Operation not permitted
error: Test failed
Test sendmsg_fs_cve failed with ret 1
Running test statx
Tests failed:  <connect> <defer> <double-poll-crash> <io_uring_enter> <io_uring_register> <link> <poll-cancel-ton> <poll-many> <poll-mshot-update> <read-write> <sq-poll-dup> <sq-poll-share> <timeout-new> <sendmsg_fs_cve>
make[1]: *** [Makefile:284: runtests] Error 1
make[1]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/liburing-liburing-2.0/test'
make: *** [Makefile:23: runtests] Error 2
```

\Here is result of running test suite over sudo
```console
+ cd liburing-liburing-2.0
+ sudo /usr/bin/make -O -j48 V=1 VERBOSE=1 runtests
make[1]: Nothing to be done for 'all'.
make[1]: Nothing to be done for 'all'.
make[1]: Nothing to be done for 'all'.
make[1]: Entering directory '/home/tkloczko/rpmbuild/BUILD/liburing-liburing-2.0/test'
Running test 232c93d07b74-test:
Running test 35fa71a030ca-test:
Test 35fa71a030ca-test timed out (may not be a failure)
Running test 500f9fbadef8-test:
Polling not supported in current dir, test skipped
Running test 7ad0e4b2f83c-test:
Running test 8a9973408177-test:
Running test 917257daa0fe-test:
Running test a0908ae19763-test:
Running test a4c0b3decb33-test:
Running test accept:
Running test accept-link:
Running test accept-reuse:
Running test accept-test:
Running test across-fork:
Running test splice:
Running test b19062a56726-test:
Running test b5837bd5311d-test:
Running test ce593a6c480a-test:
Running test close-opath:
Running test connect:
cqe 1, res 0, wanted -125
test_connect_timeout(): failed
Test connect failed with ret 1
Running test cq-full:
Running test cq-overflow:
Running test cq-peek-batch:
Running test cq-ready:
Running test cq-size:
Running test d4ae271dfaae-test:
Running test d77a67ed5f27-test:
Running test defer:
Running test double-poll-crash:
Running test eeed8b54e0df-test:
Running test eventfd:
Running test eventfd-disable:
Running test eventfd-ring:
Running test fadvise:
Running test fallocate:
Running test fc2a85cb02ef-test:
Test needs failslab/fail_futex/fail_page_alloc enabled, skipped
Running test file-register:
Skipping files not supported
Running test file-update:
Running test files-exit-hang-poll:
Running test files-exit-hang-timeout:
Running test fixed-link:
Running test fsync:
Running test io-cancel:
Test io-cancel timed out (may not be a failure)
Running test io_uring_enter:
io_uring_enter(3, 1, 0, 4294967295, (nil))
io_uring_enter(-1, 0, 0, 0, (nil))
io_uring_enter(0, 0, 0, 0, (nil))
io_uring_enter(3, 1, 0, 0, (nil))
Allocating 4096 sqes
Submitting 4096 I/Os
Done
Waiting for 4096 events
Reaping 4096 I/Os
Submitting invalid sqe index.
PASS
Running test io_uring_register:
RELIMIT_MEMLOCK: 65536 (65536)
io_uring_register(-1, 0, (nil), 0)
io_uring_register(3, 0, (nil), 0)
io_uring_register(4, 4294967295, (nil), 0)
io_uring_register(4, 0, 0x7fff266e1ed0, 1)
io_uring_register(4, 0, 0x7fff266e1ed0, 1)
io_uring_register(4, 0, 0x7fff266e1ed0, 1)
Unable to map a huge page.  Try increasing /proc/sys/vm/nr_hugepages by at least 1.
Skipping the hugepage test
reserve file-backed buffers
io_uring_register(4, 0, 0x7fff266e1ed0, 1)
io_uring_register(4, 0, 0x1c202c0, 1025)
io_uring_register(4, 0, 0x1c202c0, 1024)
allocated 17179869180 bytes of address space
io_uring_register(4, IORING_REGISTER_FILES, 0x7f07ea5e4000, 32767)...succeeded
io_uring_register(4, IORING_UNREGISTER_FILES, 0, 0)...succeeded
io_uring_submit:
        opcode: 6
        flags:  0x00000000
        fd:     4
        poll_events: 0x00000005
io_uring_register(4, 2, 0x7fff266e1ed0, 1)
PASS
Running test io_uring_setup:
io_uring_setup(0, 0x7ffe1c1235b0), flags: none, feat: none, resv: 0x00000000 0x00000000 0x00000000, sq_thread_cpu: 0
io_uring_setup(1, (nil)), flags: none, feat: none, resv: , sq_thread_cpu: 0
io_uring_setup(1, 0x7ffe1c1235b0), flags: none, feat: none, resv: 0x00000001 0x00000001 0x00000001, sq_thread_cpu: 0
io_uring_setup(1, 0x7ffe1c1235b0), flags: 0xffffffff, feat: none, resv: 0x00000000 0x00000000 0x00000000, sq_thread_cpu: 0
io_uring_setup(1, 0x7ffe1c1235b0), flags: IORING_SETUP_SQ_AFF, feat: none, resv: 0x00000000 0x00000000 0x00000000, sq_thread_cpu: 0
io_uring_setup(1, 0x7ffe1c1235b0), flags: IORING_SETUP_SQPOLL|IORING_SETUP_SQ_AFF, feat: none, resv: 0x00000000 0x00000000 0x00000000, sq_thread_cpu: 48
PASS
Running test iopoll:
File/device/fs doesn't support polled IO
Running test lfs-openat:
Running test lfs-openat-write:
Running test link:
wrong CQE order, got 1, expected 0
test_link_fail_ordering last failed
Test link failed with ret 1
Running test link-timeout:
Running test link_drain:
Running test madvise:
Running test nop:
Running test nop-all-sizes:
Running test open-close:
Running test openat2:
Running test personality:
Running test pipe-eof:
Running test pipe-reuse:
Running test poll:
Running test poll-cancel:
Running test poll-cancel-ton:
Running test poll-link:
Running test poll-many:
Running test poll-mshot-update:
submitted -16, 392
poll-many failed
Test poll-mshot-update failed with ret 1
Running test poll-ring:
Running test poll-v-poll:
Running test probe:
Running test read-write:
Running test register-restrictions:
Running test rename:
Running test ring-leak:
Running test ring-leak2:
Running test rw_merge_test:
Running test self:
Running test send_recv:
Running test send_recvmsg:
Running test shared-wq:
Running test short-read:
Running test shutdown:
Running test sigfd-deadlock:
Running test socket-rw:
Running test socket-rw-eagain:
Running test sq-full:
Running test sq-poll-dup:
Running test sq-poll-kthread:
Running test sq-poll-share:
Running test sqpoll-disable-exit:
Running test sqpoll-exit-hang:
./runtests.sh: line 66: 3367315 Killed                  timeout -s INT -k $TIMEOUT $TIMEOUT ./$test_name $dev
Test sqpoll-exit-hang failed with ret 137
Running test sqpoll-sleep:
Running test sq-space_left:
Running test stdout:
This is a pipe test
This is a fixed pipe test
Running test submit-reuse:
Running test teardowns:
Running test thread-exit:
Running test timeout:
Running test timeout-new:
Running test timeout-overflow:
Skipping
Running test unlink:
Running test wakeup-hang:
Running test sendmsg_fs_cve:
Running test statx:
Tests failed:  <connect> <link> <poll-mshot-update> <sqpoll-exit-hang>
make[1]: *** [Makefile:284: runtests] Error 1
make[1]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/liburing-liburing-2.0/test'
make: *** [Makefile:23: runtests] Error 2
```

Test suite also failed on kernel 5.10.0

```sh
$ uname -r
5.10.0-kali7-amd64
```

Here is the result of running test suite

```
$ sudo make runtests
make[1]: Entering directory '/home/d4nnylee/Documents/liburing/src'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/d4nnylee/Documents/liburing/src'
make[1]: Entering directory '/home/d4nnylee/Documents/liburing/test'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/d4nnylee/Documents/liburing/test'
make[1]: Entering directory '/home/d4nnylee/Documents/liburing/examples'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/d4nnylee/Documents/liburing/examples'
make[1]: Entering directory '/home/d4nnylee/Documents/liburing/test'
Running test 232c93d07b74-test:
Running test 35fa71a030ca-test:
Test 35fa71a030ca-test timed out (may not be a failure)
Running test 500f9fbadef8-test:
Running test 7ad0e4b2f83c-test:
Running test 8a9973408177-test:
Running test 917257daa0fe-test:
Running test a0908ae19763-test:
Running test a4c0b3decb33-test:
Running test accept:
Running test accept-link:
Running test accept-reuse:
Running test accept-test:
Running test across-fork:
Running test splice:
Running test b19062a56726-test:
Running test b5837bd5311d-test:
Running test ce593a6c480a-test:
Running test close-opath:
Running test connect:
cqe 1, res 0, wanted -125
test_connect_timeout(): failed
Test connect failed with ret 1
Running test cq-full:
Running test cq-overflow:
Running test cq-peek-batch:
Running test cq-ready:
Running test cq-size:
Running test d4ae271dfaae-test:
Running test d77a67ed5f27-test:
Running test defer:
Running test double-poll-crash:
Running test eeed8b54e0df-test:
Running test eventfd:
Running test eventfd-disable:
Running test eventfd-ring:
Running test fadvise:
Running test fallocate:
Running test fc2a85cb02ef-test:
Test needs failslab/fail_futex/fail_page_alloc enabled, skipped
Running test file-register:
Skipping files not supported
Running test file-update:
Running test files-exit-hang-poll:
Running test files-exit-hang-timeout:
Running test fixed-link:
Running test fsync:
Running test io-cancel:
Test io-cancel timed out (may not be a failure)
Running test io_uring_enter:
io_uring_enter(3, 1, 0, 4294967295, (nil))
io_uring_enter(-1, 0, 0, 0, (nil))
io_uring_enter(0, 0, 0, 0, (nil))
io_uring_enter(3, 1, 0, 0, (nil))
Allocating 4096 sqes
Submitting 4096 I/Os
Done
Waiting for 4096 events
Reaping 4096 I/Os
Submitting invalid sqe index.
PASS
Running test io_uring_register:
RELIMIT_MEMLOCK: 1554681344 (1554681344)
io_uring_register(-1, 0, (nil), 0)
io_uring_register(3, 0, (nil), 0)
io_uring_register(4, 4294967295, (nil), 0)
io_uring_register(4, 0, 0x7fff03a7da60, 1)
io_uring_register(4, 0, 0x7fff03a7da60, 1)
io_uring_register(4, 0, 0x7fff03a7da60, 1)
Unable to map a huge page.  Try increasing /proc/sys/vm/nr_hugepages by at least 1.
Skipping the hugepage test
reserve file-backed buffers
io_uring_register(4, 0, 0x7fff03a7da60, 1)
io_uring_register(4, 0, 0x558177adc6c0, 1025)
io_uring_register(4, 0, 0x558177adc6c0, 1024)
Not enough memory for this test, skipping
io_uring_submit:
	opcode: 6
	flags:  0x00000000
	fd:     4
	poll_events: 0x00000005
io_uring_register(4, 2, 0x7fff03a7d99c, 1)
PASS
Running test io_uring_setup:
io_uring_setup(0, 0x7fff223e7950), flags: none, feat: none, resv: 0x00000000 0x00000000 0x00000000, sq_thread_cpu: 0
io_uring_setup(1, (nil)), flags: none, feat: none, resv: , sq_thread_cpu: 0
io_uring_setup(1, 0x7fff223e7950), flags: none, feat: none, resv: 0x00000001 0x00000001 0x00000001, sq_thread_cpu: 0
io_uring_setup(1, 0x7fff223e7950), flags: 0xffffffff, feat: none, resv: 0x00000000 0x00000000 0x00000000, sq_thread_cpu: 0
io_uring_setup(1, 0x7fff223e7950), flags: IORING_SETUP_SQ_AFF, feat: none, resv: 0x00000000 0x00000000 0x00000000, sq_thread_cpu: 0
io_uring_setup(1, 0x7fff223e7950), flags: IORING_SETUP_SQPOLL|IORING_SETUP_SQ_AFF, feat: none, resv: 0x00000000 0x00000000 0x00000000, sq_thread_cpu: 4
PASS
Running test iopoll:
Running test lfs-openat:
Running test lfs-openat-write:
Running test link:
wrong CQE order, got 1, expected 0
test_link_fail_ordering last failed
Test link failed with ret 1
Running test link-timeout:
Running test link_drain:
Running test madvise:
Running test multicqes_drain:
Test multicqes_drain timed out (may not be a failure)
Running test nop:
Running test nop-all-sizes:
Running test open-close:
Running test openat2:
Running test personality:
Running test pipe-eof:
Running test pipe-reuse:
Running test poll:
Running test poll-cancel:
Running test poll-cancel-ton:
Running test poll-link:
Running test poll-many:
Running test poll-mshot-update:
submitted -16, 392
poll-many failed
Test poll-mshot-update failed with ret 1
Running test poll-ring:
Running test poll-v-poll:
Running test probe:
Running test read-write:
Running test register-restrictions:
submit: -77
test_restrictions_sqe_op failed
Test register-restrictions failed with ret 2
Running test rename:
Rename not supported, skipping
Running test ring-leak:
Running test ring-leak2:
Running test rw_merge_test:
Running test self:
Running test send_recv:
Non-registered SQPOLL not available, skipping
Running test send_recvmsg:
Running test shared-wq:
Running test short-read:
Running test shutdown:
Shutdown not supported, skipping
Running test sigfd-deadlock:
Running test socket-rw:
Running test socket-rw-eagain:
Running test sq-full:
Running test sq-poll-dup:
No SQPOLL sharing, skipping
No SQPOLL sharing, skipping
No SQPOLL sharing, skipping
Running test sq-poll-kthread:
Running test sq-poll-share:
No SQPOLL sharing, skipping
Running test sqpoll-disable-exit:
Running test sqpoll-exit-hang:
Skipping
Running test sqpoll-sleep:
Running test sq-space_left:
Running test stdout:
This is a pipe test
This is a fixed pipe test
Running test submit-reuse:
Running test teardowns:
Running test thread-exit:
Running test timeout:
Running test timeout-new:
feature IORING_FEAT_EXT_ARG not supported.
Test timeout-new failed with ret 1
Running test timeout-overflow:
Skipping
Running test unlink:
Unlink not supported, skipping
Running test wakeup-hang:
Running test sendmsg_fs_cve:
Running test statx:
Running test sq-full-cpp:
Tests failed:  <connect> <link> <poll-mshot-update> <register-restrictions> <timeout-new>
make[1]: *** [Makefile:286: runtests] Error 1
make[1]: Leaving directory '/home/d4nnylee/Documents/liburing/test'
make: *** [Makefile:23: runtests] Error 2
```

> Is there anything interesting in dmesg?
> Just repeated build with clear before build kernel messages buffer.

```
[ 2751.977726] Running test connect:
[ 2751.986213] TCP: request_sock_TCP: Possible SYN flooding on port 56622. Sending cookies.  Check SNMP counters.
```

--- a/test/connect.c
+++ b/test/connect.c
@@ -248,21 +248,29 @@ err1:
 
 static int test_connect_timeout(struct io_uring *ring)
 {
-	int connect_fd = -1, accept_fd = -1;
-	int ret;
+	int connect_fd[2] = {-1, -1};
+	int accept_fd = -1;
+	int ret, code;
 	struct sockaddr_in addr;
 	struct io_uring_sqe *sqe;
 	struct __kernel_timespec ts = {.tv_sec = 0, .tv_nsec = 100000};
 
-	connect_fd = create_socket();
-	if (connect_fd == -1)
+	connect_fd[0] = create_socket();
+	if (connect_fd[0] == -1)
 		return -1;
+
+	connect_fd[1] = create_socket();
+	if (connect_fd[1] == -1)
+		goto err;
 
 	accept_fd = create_socket();
 	if (accept_fd == -1)
 		goto err;
 
-	if (configure_connect(connect_fd, &addr) == -1)
+	if (configure_connect(connect_fd[0], &addr) == -1)
+		goto err;
+
+	if (configure_connect(connect_fd[1], &addr) == -1)
 		goto err;
 
 	ret = bind(accept_fd, (struct sockaddr*)&addr, sizeof(addr));
@@ -271,19 +279,28 @@ static int test_connect_timeout(struct io_uring *ring)
 		goto err;
 	}
 
-	ret = listen(accept_fd, 0);  // no backlog in order to block connect_fd
+	ret = listen(accept_fd, 0);  // no backlog in order to block connect_fd[1]
 	if (ret == -1) {
 		perror("listen()");
 		goto err;
 	}
 
+	// We first connect with one client socket in order to fill the accept queue.
+	ret = connect_socket(ring, connect_fd[0], &code);
+	if (ret == -1 || code != 0) {
+		fprintf(stderr, "unable to connect\n");
+		goto err;
+	}
+
+	// We do not offload completion events from listening socket on purpose. 
+	// This way we create a state where the second connect request being stalled by OS.  
 	sqe = io_uring_get_sqe(ring);
 	if (!sqe) {
 		fprintf(stderr, "unable to get sqe\n");
 		goto err;
 	}
 
-	io_uring_prep_connect(sqe, connect_fd, (struct sockaddr*)&addr, sizeof(addr));
+	io_uring_prep_connect(sqe, connect_fd[1], (struct sockaddr*)&addr, sizeof(addr));
 	sqe->user_data = 1;
 	sqe->flags |= IOSQE_IO_LINK;
 
@@ -320,12 +337,17 @@ static int test_connect_timeout(struct io_uring *ring)
 		io_uring_cqe_seen(ring, cqe);
 	}
 
-	close(connect_fd);
+	close(connect_fd[0]);
+	close(connect_fd[1]);
 	close(accept_fd);
 	return 0;
 
 err:
-	close(connect_fd);
+	if (connect_fd[0] != -1)
+		close(connect_fd[0]);
+	if (connect_fd[1] != -1)
+		close(connect_fd[1]);
+		
 	if (accept_fd != -1)
 		close(accept_fd);
 	return -1;

